tool/teleport/common/debug.go

```diff
@@ -0,0 +1,252 @@
+	req, err := http.NewRequestWithContext(ctx, method, "http://debug"+path, bytes.NewBuffer(body))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	resp, err := c.clt.Do(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return resp, nil
+}
+
+func onSetLogLevel(configPath string, level string) error {
+	ctx := context.Background()
+
+	if contains := slices.Contains(logutils.SupportedLevelsString, strings.ToUpper(level)); !contains {
+		return trace.BadParameter("%q log level not supported", level)
+	}
+
+	clt, err := newDebugServiceClient(configPath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	setMessage, err := clt.SetLogLevel(ctx, level)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	fmt.Println(setMessage)
+	return nil
+}
+
+func onGetLogLevel(configPath string) error {
+	ctx := context.Background()
+	clt, err := newDebugServiceClient(configPath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	level, err := clt.GetLogLevel(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	fmt.Printf("Current log level %q\n", level)
+	return nil
+}
+
+// supportedProfiles list of supported pprof profiles that can be collected.
+// This list is composed by runtime/pprof.Profile and http/pprof definitions.
+var supportedProfiles = map[string]struct{}{
+	"allocs":       {},
+	"block":        {},
+	"cmdline":      {},
+	"goroutine":    {},
+	"heap":         {},
+	"mutex":        {},
+	"profile":      {},
+	"threadcreate": {},
+	"trace":        {},
+}
+
+// defaultCollectProfiles defines the default profiles to be collected in case
+// none is provided.
+var defaultCollectProfiles = []string{"goroutine", "heap", "profile"}
+
+func onCollectProfile(configPath string, rawProfiles string, seconds int, out io.Writer) error {
+	ctx := context.Background()
+
+	profiles := defaultCollectProfiles
+	if rawProfiles != "" {
+		profiles = strings.Split(rawProfiles, ",")
+	}
+
+	for _, profile := range profiles {
+		if _, ok := supportedProfiles[profile]; !ok {
+			return trace.BadParameter("%q profile not supported", profile)
+		}
+	}
+
+	clt, err := newDebugServiceClient(configPath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var output bytes.Buffer
+	if err := createProfilesArchive(ctx, clt, &output, profiles, seconds); err != nil {
+		return trace.Wrap(err)
+	}
+
+	fmt.Fprintln(out, output.String())
+	return nil
+}
+
+// createProfileArchive collects the profiles and generate a compressed tarball
+// file.
+func createProfilesArchive(ctx context.Context, clt *debugServiceClient, buf io.Writer, profiles []string, seconds int) error {
+	gw := gzip.NewWriter(buf)
+	defer gw.Close()
+	tw := tar.NewWriter(gw)
+	defer tw.Close()
+
+	for _, profile := range profiles {
+		contents, err := clt.CollectProfile(ctx, profile, seconds)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		hd := &tar.Header{
+			Name: profile + ".pprof",
+			Size: int64(len(contents)),
+			Mode: 0600,
+		}
```

Comment:
The uncompressed file has date 1970
```code
-rw-------. 1 root root     46421 Jan  1  1970 heap.pprof
```
Not the end of the world but it would be nice to have a more accurate time.

Reply:
This seems to be specific to using GNU tar.

[From `archive/tar` docs:](https://pkg.go.dev/archive/tar#Header)

> // If the Format is unspecified, then Writer.WriteHeader rounds ModTime
  // to the nearest second and ignores the AccessTime and ChangeTime fields.
  //
  // To use AccessTime or ChangeTime, specify the Format as PAX or GNU.
  // To use sub-second resolution, specify the Format as PAX.
  
Now, I'm not sure if we should specify the format (for example, GNU) or leave it as it is.

Reply:
I think you just need the `ModTime`?
https://github.com/gravitational/teleport/blob/6b5ea0d59c04609b9b875bbc87a4be82716b1d58/tool/tctl/common/tarwriter.go#L60-L67